ggml/src/ggml-qnn/ggml-qnn.cpp

```diff
@@ -0,0 +1,4154 @@
+                      dst->nb[0], dst->nb[1], dst->nb[2], dst->nb[3]);
+    GGMLQNN_LOG_DEBUG("\n");
+}
+
+static void dump_op_info(const struct ggml_tensor * tensor) {
+    //skip sanity check of params
+    const struct ggml_tensor * src0 = tensor->src[0];
+    struct ggml_tensor       * src1 = tensor->src[1];
+    struct ggml_tensor       * dst  = const_cast<ggml_tensor *>(tensor);
+    GGMLQNN_LOG_DEBUG("op name:%s, tensor type:%s", ggml_op_name(tensor->op), ggml_type_name(tensor->type));
+    print_tensors_info(nullptr, nullptr, src0, src1, dst);
+}
+
+// =================================================================================================
+//  section-6: implementation of ggml-qnn backend
+// =================================================================================================
+//TODO: refine this function as it is a performance hotspot/bottleneck function
+static bool ggml_qnn_can_handle_op(const struct ggml_tensor * tensor) {
+    if (tensor->op == GGML_OP_NONE) {
+        return true;
+    }
+    if (ggml_is_empty(tensor) || tensor->op == GGML_OP_RESHAPE
+    || tensor->op == GGML_OP_TRANSPOSE || tensor->op == GGML_OP_VIEW
+    || tensor->op == GGML_OP_PERMUTE) {
+        return false;
+    }
+
+    //TODO: support other op
+    bool supported_op = ((tensor->op == GGML_OP_ADD) || (tensor->op == GGML_OP_MUL_MAT)
+            || (tensor->op == GGML_OP_MUL));
+    if (!supported_op) {
+        return false;
+    }
+
+    struct ggml_tensor * src0 = tensor->src[0];
+    struct ggml_tensor * src1 = tensor->src[1];
+
+    const int64_t ne00 = tensor->src[0]->ne[0];
+    const int64_t ne01 = tensor->src[0]->ne[1];
+
+    const int64_t ne10 = tensor->src[1]->ne[0];
+    const int64_t ne11 = tensor->src[1]->ne[1];
+
+    const int64_t ne0 = tensor->ne[0];
+    const int64_t ne1 = tensor->ne[1];
+
+    const uint32_t src0_rank = ggml_get_tensor_rank(src0);
+    const uint32_t src1_rank = ggml_get_tensor_rank(src1);
+
+    if (tensor->op == GGML_OP_ADD) {
+        //dump_tensors_info(tensor);
+        if (!ggml_are_same_shape(src0, src1)) {
+            return false;
+        }
+        if (ne00 < 32)
+            return false;
+        return (src0->type == GGML_TYPE_F32 || src0->type == GGML_TYPE_F16)
+               && (src1->type == GGML_TYPE_F32 || src1->type == GGML_TYPE_F16);
+    }
+
+    if (tensor->op == GGML_OP_MUL_MAT) {
+        if (src0_rank != src1_rank) // make QNN SDK happy
+            return false;
+        if (src0_rank < 2) // QNN's limitation, make QNN SDK happy
+            return false;
+        if (src0_rank > 3) //TODO: 4D matrix
+            return false;
+        if ((src1->ne[2] != src0->ne[2]) || (src1->ne[3] != src0->ne[3])) // make QNN SDK happy
+            return false;
+
+        //TODO: support more data type in func ggml_qnn_mul_mat(...)
+        //src0: q4_0, q6_k, ...
+        //src1: f32
+        //dst : f32
+        return  (src0->type == GGML_TYPE_F32 || src0->type == GGML_TYPE_F16)
```

Comment:
Noticed you posted your benchmark results with a quantized model. Wanna shout that out again:
For `MUL_MAT` operations, quantized tensors are marked as unsupported. This means your results may have been generated by the CPU backend rather than the NPU.